async_handler_test.go

```diff
@@ -0,0 +1,806 @@
+		mockCompanyCache.On("BatchDelete", mock.Anything, mock.MatchedBy(func(keys []string) bool {
+			if len(keys) != len(company.Keys) {
+				return false
+			}
+			// Check that all expected cache keys are present
+			expectedKeys := make(map[string]bool)
+			for key, value := range company.Keys {
+				cacheKey := resourceKeyToCacheKey(cacheKeyPrefixCompany, key, value)
+				expectedKeys[cacheKey] = true
+			}
+			for _, key := range keys {
+				if !expectedKeys[key] {
+					return false
+				}
+			}
+			return true
+		})).Return(nil)
+
+		// Send delete messages
+		ctx := context.Background()
+		for i := 0; i < config.BatchSize; i++ {
+			message := &schematicdatastreamws.DataStreamResp{
+				EntityType:  string(schematicdatastreamws.EntityTypeCompany),
+				MessageType: schematicdatastreamws.MessageTypeDelete,
+				Data:        companyData,
+			}
+
+			err := handler.HandleMessage(ctx, message)
+			assert.NoError(t, err)
+		}
+
+		// Wait for batch processing
+		time.Sleep(100 * time.Millisecond)
+
+		mockCompanyCache.AssertExpectations(t)
+	})
+}
+
+func TestAsyncReplicatorMessageHandler_UserBatchProcessing(t *testing.T) {
+	t.Run("processes user messages in batches", func(t *testing.T) {
+		logger := NewSchematicLogger()
+		mockCompanyCache := NewMockBatchCacheProvider[*rulesengine.Company]()
+		mockUserCache := NewMockBatchCacheProvider[*rulesengine.User]()
+		mockFlagCache := NewMockCacheProvider[*rulesengine.Flag]()
+		config := createTestAsyncConfig()
+
+		handler := NewAsyncReplicatorMessageHandler(
+			mockCompanyCache,
+			mockUserCache,
+			mockFlagCache,
+			logger,
+			5*time.Minute,
+			config,
+		)
+		defer func() {
+			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+			defer cancel()
+			handler.Shutdown(ctx)
+		}()
+
+		user := createAsyncTestUser(t)
+		userData, err := json.Marshal(user)
+		require.NoError(t, err)
+
+		// Set up expectation for batch operation
+		// Use mock.MatchedBy to handle batch items
+		mockUserCache.On("BatchSet", mock.Anything, mock.MatchedBy(func(items map[string]*rulesengine.User) bool {
+			if len(items) != len(user.Keys) {
+				return false
+			}
+			// Check that all expected cache keys are present with correct user
+			for key, value := range user.Keys {
+				cacheKey := resourceKeyToCacheKey(cacheKeyPrefixUser, key, value)
+				if u, exists := items[cacheKey]; !exists || u.ID != user.ID {
+					return false
+				}
+			}
+			return true
+		}), 5*time.Minute).Return(nil)
+
+		// Send multiple messages to trigger batching
+		ctx := context.Background()
+		for i := 0; i < config.BatchSize; i++ {
+			message := &schematicdatastreamws.DataStreamResp{
+				EntityType:  string(schematicdatastreamws.EntityTypeUser),
+				MessageType: schematicdatastreamws.MessageTypeFull,
+				Data:        userData,
+			}
+
+			err := handler.HandleMessage(ctx, message)
+			assert.NoError(t, err)
+		}
+
+		// Wait for batch processing
+		time.Sleep(100 * time.Millisecond)
```

Comment:
I am not going to block this PR, but we could perhaps invert this by giving the mock service a channel

```go
assert.NoError(t, <-mockUserCache.Done())  // done or timeout
```